hazelcast/src/main/java/com/hazelcast/spi/discovery/impl/DiscoveryServicePropertiesBuilder.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (c) 2008-2018, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.spi.discovery.impl;
+
+import com.hazelcast.config.InvalidConfigurationException;
+import com.hazelcast.config.properties.PropertyDefinition;
+import com.hazelcast.config.properties.ValidationException;
+import com.hazelcast.config.properties.ValueValidator;
+import com.hazelcast.core.TypeConverter;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import static com.hazelcast.util.MapUtil.createHashMap;
+
+/**
+ * Validates, verifies, and maps properties with the given definitions.
+ */
+class DiscoveryServicePropertiesBuilder {
+
+    /**
+     * Validates, verifies, and maps {@code properties} with the given {@code propertyDefinitions}.
+     *
+     * @param properties          properties from the Hazelcast node configuration (from the service-discovery section)
+     * @param propertyDefinitions property definitions specific for the given
+     * {@link com.hazelcast.spi.discovery.DiscoveryStrategy}
+     * @return mapped properties
+     * @throws InvalidConfigurationException if the the required properties are not satisfied or any property is not not
+     * applicable to the given definitions
+     * @throws ValidationException           if any property is invalid
+     */
+    Map<String, Comparable> buildProperties(Map<String, Comparable> properties,
+                                            Collection<PropertyDefinition> propertyDefinitions) {
+        if (propertyDefinitions == null) {
```

Comment:
why pass a null collection in the first place? In most cases it is best to convert null to an empty set so you don't need to deal with null cases.

Reply:
Yes, I completely agree. It was copied from what was before. 

I've removed it from here and added to the DefaultDiscoveryService class. Also added the method: nullToEmpty() to CollectionUtils. PTAL.